File: app/src/main/java/io/sweers/catchup/data/CatchUpItem.java
/*
 * Copyright (c) 2017 Zac Sweers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.sweers.catchup.data;

import android.support.annotation.Nullable;
import android.support.v4.util.Pair;
import com.google.auto.value.AutoValue;
import org.threeten.bp.Instant;

@AutoValue
public abstract class CatchUpItem {

  public abstract CharSequence title();

  @Nullable public abstract Pair<String, Integer> score();

  public abstract Instant timestamp();

  @Nullable public abstract String tag();

  public abstract CharSequence author();

  public abstract CharSequence source();

  public abstract int commentCount();

  public abstract boolean hideComments();

  public abstract String itemClickUrl();

  public abstract String itemCommentClickUrl();

  public static Builder builder() {
    return new AutoValue_CatchUpItem.Builder();
  }

  @AutoValue.Builder
  public interface Builder {
    Builder title(CharSequence title);

    Builder score(@Nullable Pair<String, Integer> score);

    Builder timestamp(Instant timestamp);

    Builder tag(@Nullable String tag);

    Builder author(CharSequence author);

    Builder source(CharSequence source);

    Builder commentCount(int commentCount);

    Builder hideComments(boolean hideComments);

    Builder itemClickUrl(String itemClickUrl);

    Builder itemCommentClickUrl(String itemCommentClickUrl);

    CatchUpItem build();
  }
}

File: service-api/src/commonMain/kotlin/catchup/service/api/CatchUpItem.kt
/*
 * Copyright (C) 2019. Zac Sweers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package catchup.service.api

import androidx.annotation.ColorInt
import androidx.annotation.Keep
import androidx.compose.runtime.Immutable
import catchup.service.api.ContentType.HTML
import catchup.service.api.Mark.MarkType
import java.util.Objects
import kotlinx.datetime.Instant

@Keep
@Immutable
data class CatchUpItem(
  val id: Long,
  val title: String,
  val description: String? = null,
  val timestamp: Instant? = null,
  val score: Pair<String, Int>? = null,
  val tag: String? = null,
  @ColorInt val tagHintColor: Int? = null,
  val author: String? = null,
  val source: String? = null,
  /* internal */ val itemClickUrl: String? = null,
  val imageInfo: ImageInfo? = null,
  val mark: Mark? = null,
  val detailKey: String? = null,
  val serviceId: String? = null,
  val indexInResponse: Int? = null,
  val contentType: ContentType? = null, // Null indicates unset, try to infer it
) {

  val clickUrl: String?
  val markClickUrl: String?

  init {
    val markClickUrl = mark?._markClickUrl
    val itemClickUrl = itemClickUrl ?: markClickUrl
    val finalMarkClickUrl = markClickUrl?.let { if (itemClickUrl == markClickUrl) null else it }
    this.clickUrl = itemClickUrl
    this.markClickUrl = finalMarkClickUrl
  }

  companion object {
    fun fakeItems(count: Int, serviceId: String, isImage: Boolean): List<CatchUpItem> {
      return (0 until count).map { fake(it, serviceId, isImage) }
    }

    fun fake(
      index: Int,
      serviceId: String,
      isImage: Boolean,
      id: Long = Objects.hash(index, serviceId).toLong(),
    ): CatchUpItem {
      val imageInfo =
        if (isImage) {
          val url = "https://picsum.photos/seed/$index/300/300"
          ImageInfo(
            url = url,
            bestSize = 400 to 300,
            animatable = false,
            detailUrl = url,
            sourceUrl = url,
            aspectRatio = 4 / 3f,
            imageId = url,
          )
        } else {
          null
        }

      return CatchUpItem(
        id = id,
        title = "Lorem ipsum dolor sit amet",
        description =
          "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        timestamp = Instant.parse("2020-01-01T00:00:00Z"),
        score = "+" to 5,
        tag = "Tag",
        author = "Author",
        source = "Source",
        itemClickUrl = "https://example.com",
        imageInfo = imageInfo,
        mark = Mark(markType = MarkType.COMMENT, _markClickUrl = "https://example.com"),
        detailKey = "0",
        serviceId = serviceId,
        indexInResponse = index,
        contentType = HTML,
      )
    }
  }
}

val CatchUpItem.canBeSummarized: Boolean
  get() = contentType == HTML

val CatchUpItem.supportsDetail: Boolean
  get() = detailKey != null
